src/hierarchies/base_hierarchy.hpp

```diff
@@ -30,37 +32,47 @@
 class BaseHierarchy {
  protected:
   // map: data_id -> datum
-  std::unordered_map<int, Eigen::VectorXd> cluster_data;
+  std::set<int> cluster_data_idx;
   int card = 0;
-  double log_card;
+  double log_card = stan::math::NEGATIVE_INFTY;
+
+  virtual void update_summary_statistics(const Eigen::VectorXd &datum,
+                                         bool add) = 0;
 
  public:
   void add_datum(const int &id, const Eigen::VectorXd &datum) {
-    cluster_data[id] = datum;
+    auto it = cluster_data_idx.find(id);
+    assert(it == cluster_data_idx.end());
     card += 1;
     log_card = std::log(card);
     update_summary_statistics(datum, true);
+    cluster_data_idx.insert(id);
+
   }
 
-  void remove_datum(const int &id) {
-    auto it = cluster_data.find(id);
-    assert(it != cluster_data.end());
-    update_summary_statistics(cluster_data[id], false);
-    cluster_data.erase(it);
+  void remove_datum(const int &id, const Eigen::VectorXd &datum) {
+    update_summary_statistics(datum, false);
     card -= 1;
-    log_card = std::log(card);
+    log_card = (card == 0) ? stan::math::NEGATIVE_INFTY : std::log(card);
+    auto it = cluster_data_idx.find(id);
+    assert(it != cluster_data_idx.end());
+    cluster_data_idx.erase(it);
   }
 
   int get_card() const { return card; }
-  double get_log_card() const { return card; }
+  double get_log_card() const { return log_card; }
+
+  void set_card(int card_) { 
+    card = card_; 
+    log_card = std::log(card_);
+  }
```

Comment:
I know you need in the temporary hierarchy, but the very existence of this setter is kinda shaky... Is there not a better way? Like a special function/case for add_datum, or a special constructor for the class?

Reply:
I don't really mind this since basically we are changing the state of the Hierarchy to be both the unique values and the cardinality of the cluster.
If you have better ideas we can change it

Reply:
If you're treating the cardinality as part of the state, it's even worse, because you're independently changing a single portion of the state from the outside!

I'd like the hierarchies to be consistent at all times, as I believe avoiding doing so can have unintended consequences later if we forget. Maybe it's enough to add a line like this in the setter:
`cluster_data_idx = std::set<int>(card_);`

Reply:
There is no such constructor for a `std::set`, since they are containers for unique elements

Reply:
would changing the `void set_state_from_proto(const google::protobuf::Message &state_)` method to, for instance
```
void set_state((const google::protobuf::Message &val_, const int& size)
```
work for you?

Reply:
Or we can change the ClusterState message to 
```
message ClusterState {
    oneof val {
      UnivLSState univ_ls_state = 1;
      MultiLSState multi_ls_state = 2;
    }
    int32 cardinality = 3;
}
```
and remove the cardinalities from the MarginalState message

Reply:
I like the latter solution a lot, that's the way the state is supposed to be. Doing so would remove the need of set_card(), correct? If so I'm all up to it!